File: test/hardware/integration/test_forwarding.py
import cocotb
from cocotb.triggers import Timer, RisingEdge
from cocotb.clock import Clock
from pathlib import Path
from test.driver import run_hardware_test
from test.env import get_all_rtl_files

# Program Layout:
# 0x00: ADDI x1, x0, 10      (x1 = 10)
# 0x04: ADD  x2, x1, x1      (x2 = 20) -> Tests GPR Forwarding (EX->EX)
# 0x08: ADDI x3, x0, 0x40    (x3 = 0x40)
# 0x0C: CSRRW x0, mtvec, x3  (mtvec = 0x40)
# 0x10: ECALL                (Trap to 0x40) -> Tests CSR Forwarding (mtvec)
# ...
# 0x40: ADDI x4, x0, 0x80    (x4 = 0x80)
# 0x44: CSRRW x0, mepc, x4   (mepc = 0x80)
# 0x48: MRET                 (Return to 0x80) -> Tests CSR Forwarding (mepc)
# ...
# 0x80: ADDI x10, x0, 1      (Success)
# 0x84: EBREAK

PROGRAM_SIZE = 256
PROGRAM = ["00000013"] * PROGRAM_SIZE # Initialize with NOPs

# Instructions
PROGRAM[0] = "00a00093" # 0x00: ADDI x1, x0, 10
PROGRAM[1] = "00108133" # 0x04: ADD x2, x1, x1
PROGRAM[2] = "04000193" # 0x08: ADDI x3, x0, 0x40
PROGRAM[3] = "30519073" # 0x0C: CSRRW x0, mtvec, x3
PROGRAM[4] = "00000073" # 0x10: ECALL

# Handler at 0x40 (Index 16)
PROGRAM[16] = "08000213" # 0x40: ADDI x4, x0, 0x80
PROGRAM[17] = "34121073" # 0x44: CSRRW x0, mepc, x4
PROGRAM[18] = "30200073" # 0x48: MRET

# Target at 0x80 (Index 32)
PROGRAM[32] = "00100513" # 0x80: ADDI x10, x0, 1
PROGRAM[33] = "00100073" # 0x84: EBREAK

def create_hex_file(filename="program.hex"):
    with open(filename, "w") as f:
        for instr in PROGRAM:
            f.write(f"{instr}\n")

@cocotb.test()
async def test_forwarding_program(dut):
    create_hex_file("program.hex")
    clock = Clock(dut.clk, 10, units="ns")
    cocotb.start_soon(clock.start())
    dut.rst_n.value = 0
    await Timer(20, units="ns")
    dut.rst_n.value = 1
    
    # Run simulation
    for i in range(200):
        await RisingEdge(dut.clk)
        try:
            pc = dut.u_tile_0.u_core.u_backend.id_ex_program_counter.value.integer
            x2 = dut.u_tile_0.u_core.u_backend.u_regfile.registers[2].value.integer
            x10 = dut.u_tile_0.u_core.u_backend.u_regfile.registers[10].value.integer
            
            dut._log.info(f"Cycle {i}: PC={hex(pc)}, x2={x2}, x10={x10}")
            
            if pc == 0x84: # EBREAK
                break
        except Exception:
            pass
            
    for _ in range(5):
        await RisingEdge(dut.clk)

    # Check GPR Forwarding Result
    x2 = dut.u_tile_0.u_core.u_backend.u_regfile.registers[2].value.integer
    assert x2 == 20, f"GPR Forwarding Failed: x2 should be 20, got {x2}"
    
    # Check CSR Forwarding Result (Reached end of program)
    x10 = dut.u_tile_0.u_core.u_backend.u_regfile.registers[10].value.integer
    assert x10 == 1, f"CSR Forwarding Failed: x10 should be 1, got {x10}"

def test_forwarding():
    run_hardware_test(
        module_name=Path(__file__).stem,
        toplevel="chip_top",
        verilog_sources=get_all_rtl_files()
    )

File: test/hardware/integration/test_forwarding.cpp
// Test: Forwarding Integration Test
// Tests both GPR and CSR forwarding paths:
// - GPR Forwarding (EX->EX): ADDI x1=10, ADD x2=x1+x1 (x2=20)
// - CSR Forwarding: ECALL -> mtvec handler -> MRET -> mepc target
// Program Layout:
// 0x00: ADDI x1, x0, 10      (x1 = 10)
// 0x04: ADD  x2, x1, x1      (x2 = 20) -> Tests GPR Forwarding
// 0x08: ADDI x3, x0, 0x40    (x3 = 0x40)
// 0x0C: CSRRW x0, mtvec, x3  (mtvec = 0x40)
// 0x10: ECALL                (Trap to 0x40)
// ...
// 0x40: ADDI x4, x0, 0x80    (x4 = 0x80)
// 0x44: CSRRW x0, mepc, x4   (mepc = 0x80)
// 0x48: MRET                 (Return to 0x80)
// ...
// 0x80: ADDI x10, x0, 1      (Success marker)
// 0x84: EBREAK

#include "../common/tb_base.h"
#include <Vchip_top.h>

class ChipTopTestbench : public ClockedTestbench<Vchip_top> {
public:
    void set_clk(uint8_t value) override {
        dut->clk = value;
    }

    void load_program(const std::vector<uint32_t>& program) {
        for (size_t i = 0; i < program.size(); i++) {
            dut->rootp->chip_top__DOT__u_main_memory__DOT__memory[i] = program[i];
        }
    }

    uint32_t read_register(int reg_idx) {
        if (reg_idx < 0 || reg_idx >= 32) return 0;
        return dut->rootp->chip_top__DOT__u_tile_0__DOT__u_core__DOT__u_backend__DOT__u_regfile__DOT__registers[reg_idx];
    }

    uint32_t get_pc_ex() {
        return dut->rootp->chip_top__DOT__u_tile_0__DOT__u_core__DOT__u_backend__DOT__id_ex_program_counter;
    }
};

int main(int argc, char** argv) {
    Verilated::commandArgs(argc, argv);
    
    ChipTopTestbench tb;
    tb.open_trace("dump.vcd");

    // Initialize program with NOPs
    std::vector<uint32_t> program(256, 0x00000013); // 256 NOPs

    // Main instructions
    program[0] = 0x00a00093;  // 0x00: ADDI x1, x0, 10
    program[1] = 0x00108133;  // 0x04: ADD x2, x1, x1
    program[2] = 0x04000193;  // 0x08: ADDI x3, x0, 0x40
    program[3] = 0x30519073;  // 0x0C: CSRRW x0, mtvec, x3
    program[4] = 0x00000073;  // 0x10: ECALL

    // Handler at 0x40 (Index 16)
    program[16] = 0x08000213; // 0x40: ADDI x4, x0, 0x80
    program[17] = 0x34121073; // 0x44: CSRRW x0, mepc, x4
    program[18] = 0x30200073; // 0x48: MRET

    // Target at 0x80 (Index 32)
    program[32] = 0x00100513; // 0x80: ADDI x10, x0, 1
    program[33] = 0x00100073; // 0x84: EBREAK

    tb.reset();
    tb.load_program(program);

    // Run until EBREAK
    bool ebreak_reached = false;
    for (int cycles = 0; cycles < 200; cycles++) {
        tb.tick();
        
        uint32_t pc_ex = tb.get_pc_ex();
        if (pc_ex == 0x84) { // EBREAK instruction address
            ebreak_reached = true;
            for (int i = 0; i < 5; i++) tb.tick();
            break;
        }
    }

    TB_ASSERT_EQ(ebreak_reached, true, "EBREAK should be reached");

    // Check GPR Forwarding Result
    uint32_t x2 = tb.read_register(2);
    TB_ASSERT_EQ(x2, 20, "GPR Forwarding: x2 should be 20");

    // Check CSR Forwarding Result (Reached end of program)
    uint32_t x10 = tb.read_register(10);
    TB_ASSERT_EQ(x10, 1, "CSR Forwarding: x10 should be 1");

    return 0;
}
